xla/tests/onednn_softmax_test.cc

```diff
@@ -99,6 +106,46 @@ TEST_F(OneDnnSoftmaxTest, SoftmaxFP32) {
     )";
 
   EXPECT_TRUE(RunAndCompare(hlo_string, ErrorSpec{1e-4, 1e-4}));
+  MatchOptimizedHlo(hlo_string, onednn_softmax_);
+}
+
+TEST_F(OneDnnSoftmaxTest, SoftmaxFP32_0) {
+  const std::string hlo_string = R"(
+        HloModule jit_softmax, entry_computation_layout={(f32[3,1,1]{2,1,0})->f32[3,1,1]{2,1,0}}
+        region_0.5 {
+          Arg_0.6 = f32[] parameter(0)
+          Arg_1.7 = f32[] parameter(1)
+          ROOT maximum.8 = f32[] maximum(Arg_0.6, Arg_1.7)
+        }
+        region_1.17 {
+          Arg_0.18 = f32[] parameter(0)
+          Arg_1.19 = f32[] parameter(1)
+          ROOT add.20 = f32[] add(Arg_0.18, Arg_1.19)
+        }
+        ENTRY main.27 {
+          Arg_0.1 = f32[3,1,1]{2,1,0} parameter(0)
+          constant.4 = f32[] constant(-inf)
+          reduce.9 = f32[1,1]{1,0} reduce(Arg_0.1, constant.4), dimensions={0}, to_apply=region_0.5
+          constant.2 = f32[1,1]{1,0} constant({ {-inf} })
+          maximum.10 = f32[1,1]{1,0} maximum(reduce.9, constant.2)
+          reshape.11 = f32[1,1,1]{2,1,0} reshape(maximum.10)
+          broadcast.12 = f32[1,1,1]{2,1,0} broadcast(reshape.11), dimensions={0,1,2}
+          reshape.13 = f32[1,1]{1,0} reshape(broadcast.12)
+          broadcast.14 = f32[3,1,1]{2,1,0} broadcast(reshape.13), dimensions={1,2}
+          subtract.15 = f32[3,1,1]{2,1,0} subtract(Arg_0.1, broadcast.14)
+          exponential.16 = f32[3,1,1]{2,1,0} exponential(subtract.15)
+          constant.3 = f32[] constant(0)
+          reduce.21 = f32[1,1]{1,0} reduce(exponential.16, constant.3), dimensions={0}, to_apply=region_1.17
+          reshape.22 = f32[1,1,1]{2,1,0} reshape(reduce.21)
+          broadcast.23 = f32[1,1,1]{2,1,0} broadcast(reshape.22), dimensions={0,1,2}
+          reshape.24 = f32[1,1]{1,0} reshape(broadcast.23)
+          broadcast.25 = f32[3,1,1]{2,1,0} broadcast(reshape.24), dimensions={1,2}
+          ROOT divide.26 = f32[3,1,1]{2,1,0} divide(exponential.16, broadcast.25)
+        }
+    )";
+
+  EXPECT_TRUE(RunAndCompare(hlo_string, ErrorSpec{1e-4, 1e-4}));
```

Comment:
For every test , we check the numerical correctness as well as the custom call target to verify if the pattern matched via filecheck. Though it might be redundant to add check for just rewrite pass, I included them for every test, would that work ?

Reply:
Thanks---but indeed, done as is, it is a bit redundant!

What I wanted to suggest is that you write tests that do not exercise the full pipeline, but only test the pass. I think there are three types of tests you want to have:

1. **Pass-specific tests**: tests that check that your pass itself matches what it should. Those tests should call into the pass and make sure that it rewrites the right patterns into the right expected custom call; 
2. **Pipeline test**: test that checks that the compiler pipeline calls your pass appropriately. This test should call one full compilation loop and make sure that the relevant pass is called;
3. **Numerics tests**: tests that your custom-calls produce the right numerics vs a reference implementation. These tests should execute already optimized code (post pattern matching) and compare them against a reference implementation corresponding to what should have been matched.

Today, (1), (2) and (3) are sort of all bundled together. I claim that it makes sense to separate the 3 types of tests into separate tests, for several reasons:

* Bundling all three types of tests together mean that we always pay more runtime cost than necessary for the same coverage.
* In general, (1) and (2) are tests that shouldn't rely on specific hardware, but (3) may require the test to run on a specific platform. Splitting the test targets means that we can save resources by default, and only specialise on the platform where necessary.
* Testing (2) by matching the existence of a custom call at the end of the pipeline is somewhat brittle and intrusive. Someone adding a pass that causes your input HLO to look different at the start of your pass might need to come back to fix these tests, even though they might have not broken anything. This means that we want to (at the very least) have as few of this type of tests as possible to minimize churns. In your case, I think that checking that the optimized HLO has gone through the right pass by looking at the module metadata would be enough.

The request here would thus be to split tests into these three categories. Does that make more sense?

Reply:
Makes sense and thank you for clarification. I modified the test to just check the rewrite pass.